### ISSUE TITLE
IDISP007 conditional instantiation of optional dependencies

### ISSUE THREAD

**Issue by kuzyaka** [DotNetAnalyzers/IDisposableAnalyzers#85]
```
public class Foo : IDisposable
{
		private readonly Mutex dependency;

		public Foo(Mutex dependency = null)
		{
			this.dependency = dependency;
			dependency = dependency ?? new Mutex();
		}
		
		public void Dispose()
		{
			dependency.Dispose(); // IDISP007
		}
}

// usage with false positive IDISP007
using (var foo = new Foo()) 
{
}
```

Dependency may be created in constructor, in that case IDISP007 warn is incorrect. Analyzer should scan usages to surely diagnose this case.

**Comment by JohanLarsson** [comment]
```cs
namespace RoslynSandbox
{
    using System;
    using System.Threading;

    public class Foo : IDisposable
    {
        private readonly Mutex dependency; // IDISP008

        public Foo(Mutex dependency = null)
        {
            this.dependency = dependency ?? new Mutex();
        }

        public void Dispose()
        {
            dependency.Dispose(); // IDISP007
        }
    }
}
```

For the above code I think the warnings are correct.

- IDISP008 because we assign `dependency` with either the injected ctor parameter or `new Mutex()` and hence we don't know if we are responsible for disposing it.
- IDISP007 because `dependency` is potentially injected.

**Comment by kuzyaka** [comment]
As far as I understand, this is OK for now due to the complexity of implementing more advanced diagnostics that would take into account the external uses of the diagnosed classes. I understand this limitation, it is a pity that it leads to a pretty large amount of false positives in my use case.

**Comment by JohanLarsson** [comment]
I don't agree the warnings above are false positives, in fact we put a lot of work in to warn in these cases.
As I see it the design is not clear. An alternative design could be:

```cs
namespace RoslynSandbox
{
    using System;
    using System.Threading;

    public class Foo : IDisposable
    {
        private readonly Mutex injected;
        private readonly Mutex created;

        public Foo(Mutex dependency = null)
        {
            this.injected = dependency;
            if (dependency == null)
            {
                this.created = new Mutex();
            }
        }

        private Mutex Effective => this.injected ?? this.created;

        public void Dispose()
        {
            this.created?.Dispose();
        }
    }
}
```

Above is perhaps not pretty but it is explicit about ownership.

**Comment by JohanLarsson** [comment]
This is probably a nicer design:
```cs
public class Foo : IDisposable
{
    private MutexWrapper mutex;

    public Foo(Mutex dependency = null)
    {
        this.mutex = new MutexWrapper(dependency);
    }

    public void Dispose()
    {
        this.mutex.Dispose();
    }

    private class MutexWrapper : IDisposable
    {
        private readonly Mutex injected;
        private readonly Mutex created;

        public MutexWrapper(Mutex dependency = null)
        {
            this.injected = dependency;
            if (dependency == null)
            {
                this.created = new Mutex();
            }
        }

        public Mutex Effective => this.injected ?? this.created;

        public void Dispose()
        {
            this.created?.Dispose();
        }
    }
}
```

**Comment by kuzyaka** [comment]
Thank you for workaround. I guess in this case, the IDisposableAnalyzer user will most likely mute the warning, instead of rewriting to the form convenient for the analyzer

**Comment by JohanLarsson** [comment]
> As far as I understand, this is OK for now due to the complexity of implementing more advanced diagnostics that would take into account the external uses of the diagnosed classes.

We can't do this, we could do it for the current solution but that does not mean it is used as a library. In short we can say anything about a mutable member that is not private.

>  I guess in this case, the IDisposableAnalyzer user will most likely mute the warning.

Suppressing it is fine of course, while a pragma is not pretty it provides some clarity in situations like these.

> instead of rewriting to the form convenient for the analyzer

It is not really a convenient for for the analyzer, it is rewriting it so that it can be reasoned about at all.

**Comment by kuzyaka** [comment]
> We can't do this, we could do it for the current solution but that does not mean it is used as a library. In short we can say anything about a mutable member that is not private.

mutable member is "mutex" field, it is private

**Comment by JohanLarsson** [comment]
Yes true but it is assigned with the constructor parameter and the constructor is public meaning we will not know where the value comes from and if it is ok to dispose it.

**Comment by kuzyaka** [comment]
Your arguments are clear. Thank you
